integration/test/Testlib/ModService.hs

```diff
@@ -281,63 +284,119 @@ startBackend resource overrides services = do
+
+        waitForService <-
+          appToIOKleisli
+            \srv ->
+              retryRequestUntil
+                do checkServiceIsUp srv
+                do show srv
+
+        liftIO do waitForService original
+        ensureReachable resource.berDomain
+
+        -- the service has come up, we can safely put the lock in
+        liftIO do putMVar lock ()
+
+        -- return the running service
+        pure running
+
+      -- run a service to completion
+      waitService :: RunningService -> App ExitCode
+      waitService = liftIO . waitForProcess . processHandle
+
+      -- terminate a process
+      closeService :: RunningService -> IO ()
+      closeService runningProcess = do
+        -- acquire handles for the running process
+        -- (stdout, stderr, process handle)
+        let hdls = extractHandles runningProcess
+            ph = processHandle runningProcess
+            path = configPath runningProcess
+        -- try killing the process while cleaning up the handlers
+        cleanupProcess hdls
+        timeout 50000 (waitForProcess ph) >>= \case
+          Just _ -> pure ()
+          Nothing -> do
+            putStrLn "Process has not terminated before timing out the first time"
+            timeout 100000 (waitForProcess ph) >>= \case
+              Just _ -> pure ()
+              Nothing -> do
+                putStrLn "Process has not terminated before timing out the second time"
+                mPid <- getPid ph
+                for_ mPid (signalProcess killProcess)
+                void $ waitForProcess ph
+        whenM (doesFileExist path) $ removeFile path
+        whenM (doesDirectoryExist path) $ removeDirectoryRecursive path
+
+      -- bracketing services with a lock
+      -- the three arguments work as follows
+      -- - acquire a lock by starting up a service and waiting for it to come up
+      -- - kill the process (upon completion or error)
+      -- - wait for either
+      --   - the lock to be released (asking for the service to be terminated)
+      --   - the process to run to completion
+      bracketServiceWithLock lock ser =
+        UnliftIO.bracket
+          do initService lock ser
+          do liftIO . closeService
+          \running -> do
+            waitServiceIO <- appToIO do
+              waitService running
+            liftIO do
+              -- racing termination from the outside
+              -- and the actual service
+              race
+                -- wait for the lock to be freed again (meaning release the process)
+                do putMVar lock ()
+                -- run the process to finish
+                do waitServiceIO
+              >>= liftIO . either
+                do const (putStrLn "process has been terminated")
+                \code -> putStrLn do
+                  "process finished with exit code " <> show code
+
+      modifyEnv env = env {serviceMap = Map.insert domain serviceMap env.serviceMap}
+
+  lift $ local modifyEnv do
```

Comment:
I'm actually not sure whether I've done this right; what do I have to do with the continuation here?